src/middlewared/middlewared/plugins/iscsi.py

```diff
@@ -833,34 +834,22 @@ async def disk_choices(self):
         """
         diskchoices = {}
 
-        zvol_query_filters = [('type', '=', 'DISK')]
-
-        used_zvols = [
-            i['path'] for i in (await self.query(zvol_query_filters))
-        ]
-
         zvols = await self.middleware.call(
-            'pool.dataset.query', [
-                ('type', '=', 'VOLUME'),
-                ('locked', '=', False)
-            ]
-        )
-        zvol_list = []
-        for zvol in zvols:
-            zvol_name = zvol['name']
-            zvol_size = zvol['volsize']['value']
-            zvol_list.append(zvol_name)
-            key = os.path.relpath(zvol_name_to_path(zvol_name), '/dev')
-            if key not in used_zvols:
-                diskchoices[key] = f'{zvol_name} ({zvol_size})'
-
-        zfs_snaps = await self.middleware.call(
-            'zfs.snapshot.query',
-            [['dataset', 'in', [zvol['name'] for zvol in zvols]]],
-            {'select': ['name']},
+            'zfs.dataset.unlocked_zvols_fast',
+            {'additional_information': ['SIZE', 'RO', 'ATTACHMENT']}
         )
-        for snap in zfs_snaps:
-            diskchoices[os.path.relpath(zvol_name_to_path(snap['name']), '/dev')] = f'{snap["name"]} [ro]'
+
+        for zvol in zvols.values():
+            if zvol['attachment'] is not None:
```

Comment:
You could take advantage of a filterable interface here

Reply:
Same filtering goes on in either one function or another. I was returning dict so that I can get O(1) lookups for info based on dataset names. This is more important for me in future function to do same for datasets. ZFS objset kstats don't differentiate between whether something is a zvol or filesystem.